### PR TITLE
[WIP] Bb merkelized

### PR DESCRIPTION
**Pull request by DieMyst**

### CODE REVIEW

File: statemachine/merkelized-bytebuffer/src/main/scala/fluence/merkle/BinaryMerkleTree.scala [hunk 1]
```diff
@@ -0,0 +1,237 @@
...
+    chunkSize: Int,
+    ops: MerkleOperations[T],
+    storage: Storage[T]
+  )(implicit c: ClassTag[T]): BinaryMerkleTree[T] = {
+    import TreeMath._
+
+    assert(size % chunkSize == 0)
+
+    // count of leafs that mapped on byte array (can have non-zero values)
+    val mappedLeafCount = size / chunkSize + (if (size % chunkSize == 0) 0 else 1)
```
Comment (michaelvoronov) [inline statemachine/merkelized-bytebuffer/src/main/scala/fluence/merkle/BinaryMerkleTree.scala]: val mappedLeafCount = (size +chunkSize - 1) / chunkSize
is faster

File: statemachine/merkelized-bytebuffer/src/main/scala/fluence/merkle/BinaryMerkleTree.scala [hunk 2]
```diff
@@ -0,0 +1,237 @@
...
+
+    // number of leafs in ByteBuffer
+    val leafCount = {
+      val power = log2(mappedLeafCount)
+      val k = power2(power)
+      if (k >= mappedLeafCount) k
+      else power2(power + 1)
```
Comment (michaelvoronov) [inline statemachine/merkelized-bytebuffer/src/main/scala/fluence/merkle/BinaryMerkleTree.scala]: From my point of view, here it is better not to use power of two for LeafsCount since our array has a fixed size and couldn't be growed. Yes, getParentPos would be more complex, but it is possible to realize it in this case.

File: statemachine/merkelized-bytebuffer/src/main/scala/fluence/merkle/BinaryMerkleTree.scala [hunk 3]
```diff
@@ -0,0 +1,237 @@
...
+  )(implicit c: ClassTag[T]): BinaryMerkleTree[T] = {
+    import TreeMath._
+
+    assert(size % chunkSize == 0)
+
+    // count of leafs that mapped on byte array (can have non-zero values)
+    val mappedLeafCount = size / chunkSize + (if (size % chunkSize == 0) 0 else 1)
+
+    // number of leafs in ByteBuffer
+    val leafCount = {
```
Comment (michaelvoronov) [inline statemachine/merkelized-bytebuffer/src/main/scala/fluence/merkle/BinaryMerkleTree.scala]: https://graphics.stanford.edu/~seander/bithacks.html#RoundUpPowerOf2

File: statemachine/merkelized-bytebuffer/src/main/scala/fluence/merkle/BinaryMerkleTree.scala [hunk 4]
```diff
@@ -0,0 +1,237 @@
...
+      val power = log2(mappedLeafCount)
+      val k = power2(power)
+      if (k >= mappedLeafCount) k
+      else power2(power + 1)
+    }
+
+    val treeHeight = {
+      val h = log2(leafCount)
+      if (h == 0) 1
```
Comment (michaelvoronov) [inline statemachine/merkelized-bytebuffer/src/main/scala/fluence/merkle/BinaryMerkleTree.scala]: It could be less then 0 because of upper checks

File: statemachine/merkelized-bytebuffer/src/main/scala/fluence/merkle/BinaryMerkleTree.scala [hunk 5]
```diff
@@ -0,0 +1,237 @@
...
+      else h
+    }
+
+    val numberOfNodes = power2(treeHeight + 1) - 1
+
+    val arrayTree = new Array[T](numberOfNodes)
+
+    val tree = new BinaryMerkleTree(arrayTree, treeHeight, chunkSize, mappedLeafCount, ops, storage)
+    tree.recalculateAll()
```
Comment (michaelvoronov) [inline statemachine/merkelized-bytebuffer/src/main/scala/fluence/merkle/BinaryMerkleTree.scala]: For more fast wasm files loading it should be precalculated.

File: statemachine/merkelized-bytebuffer/src/main/scala/fluence/merkle/TreeMath.scala [hunk 6]
```diff
@@ -0,0 +1,15 @@
+package fluence.merkle
+
+object TreeMath {
+
+  // log2(input) for integers
+  def log2(input: Int): Int = {
+    val res = 31 - Integer.numberOfLeadingZeros(input)
```
Comment (michaelvoronov) [inline statemachine/merkelized-bytebuffer/src/main/scala/fluence/merkle/TreeMath.scala]: https://graphics.stanford.edu/~seander/bithacks.html#IntegerLogDeBruijn - O(log n) instead of current O(n)

File: statemachine/merkelized-bytebuffer/src/main/scala/fluence/merkle/storage/ByteBufferWrapper.scala [hunk 7]
```diff
@@ -0,0 +1,193 @@
...
+import java.util
+
+class ByteBufferWrapper(val bb: ByteBuffer, chunkSize: Int) extends Storage[Array[Byte]] {
+
+  private val touchedChunks = new util.BitSet(chunkSize)
+
+  def getTouched: util.BitSet = touchedChunks
+
+  override def getElements(offset: Int, length: Int): Array[Byte] = {
+    val arr = new Array[Byte](length)
```
Comment (michaelvoronov) [inline statemachine/merkelized-bytebuffer/src/main/scala/fluence/merkle/storage/ByteBufferWrapper.scala]: Since it has public put/get methods based only on position it is better to use duplicate first, like [here](https://github.com/fluencelabs/fluence/blob/master/vm/src/main/scala/fluence/vm/wasm/WasmModuleMemory.scala#L44)

File: statemachine/merkelized-bytebuffer/src/main/scala/fluence/merkle/BinaryMerkleTree.scala [hunk 8]
```diff
@@ -0,0 +1,237 @@
...
+   * @param dirtyNodes list of dirty nodes
+   *
+   * @return root hash
+   */
+  @scala.annotation.tailrec
+  private def recalculateNodes(rowSize: Int, height: Int, dirtyNodes: util.BitSet): T = {
+    var i = -1
+    var exit = true
+    val parentsRowSize = rowSize / 2
+    val parents = new util.BitSet(parentsRowSize)
```
Comment (michaelvoronov) [inline statemachine/merkelized-bytebuffer/src/main/scala/fluence/merkle/BinaryMerkleTree.scala]: Maybe use only one top-level allocated BitSet?

File: statemachine/merkelized-bytebuffer/src/main/scala/fluence/merkle/BinaryMerkleTree.scala [hunk 9]
```diff
@@ -0,0 +1,237 @@
...
+    bs.set(0, allLeafs)
+    recalculateLeafs(allLeafs, bs)
+  }
+
+  /**
+   * Gets data from storage and calculates hash.
+   *
+   * @param pos leaf position
+   */
+  private def calculateLeafHash(pos: Int): Unit = {
```
Comment (michaelvoronov) [inline statemachine/merkelized-bytebuffer/src/main/scala/fluence/merkle/BinaryMerkleTree.scala]: It seems that name of these methods should start with "update" because they not only calculates but also updates an appropriate value.

File: statemachine/merkelized-bytebuffer/src/main/scala/fluence/merkle/BinaryMerkleTree.scala [hunk 10]
```diff
@@ -0,0 +1,237 @@
...
+    val nextHeight = height - 1
+    if (nextHeight == 0) calculateRootHash()
+    else recalculateNodes(parentsRowSize, nextHeight, parents)
+  }
+
+  private def recalculateLeafs(size: Int, bits: util.BitSet): T = {
+    recalculateNodes(size, treeHeight, bits)
+  }
+
+  def showTree(): Unit = {
```
Comment (alari) [inline statemachine/merkelized-bytebuffer/src/main/scala/fluence/merkle/BinaryMerkleTree.scala]: Usually `show` refers to cats `Show` typeclass.

File: statemachine/merkelized-bytebuffer/src/main/scala/fluence/merkle/storage/ByteBufferWrapper.scala [hunk 11]
```diff
@@ -0,0 +1,193 @@
+package fluence.merkle.storage
+
+import java.nio.ByteBuffer
+import java.util
+
+class ByteBufferWrapper(val bb: ByteBuffer, chunkSize: Int) extends Storage[Array[Byte]] {
+
+  private val touchedChunks = new util.BitSet(chunkSize)
+
+  def getTouched: util.BitSet = touchedChunks
```
Comment (alari) [inline statemachine/merkelized-bytebuffer/src/main/scala/fluence/merkle/storage/ByteBufferWrapper.scala]: Maybe use scodec bits instead? Was it considered as an option?

File: project/SbtCommons.scala [hunk 12]
```diff
@@ -126,6 +126,8 @@ object SbtCommons {
 
   val rocksDb = "org.rocksdb" % "rocksdbjni" % "5.17.2"
 
+  val apacheCommons = "org.apache.commons" % "commons-lang3" % "3.9"
```
Comment (alari) [inline project/SbtCommons.scala]: is it possible not to bring this dependency into the project just to use it in `Show` function?

`StringUtils.center` is the only used function I have found.

File: statemachine/merkelized-bytebuffer/src/main/scala/fluence/merkle/BinaryMerkleTree.scala [hunk 13]
```diff
@@ -0,0 +1,237 @@
...
+import cats.syntax.show._
+import fluence.merkle.ops.{ByteMerkleOperations, MerkleOperations}
+import fluence.merkle.storage.Storage
+import org.apache.commons.lang3.StringUtils
+import scodec.bits.ByteVector
+
+import scala.language.higherKinds
+import scala.reflect.ClassTag
+
+class BinaryMerkleTree[T: Show] private (
```
Comment (alari) [inline statemachine/merkelized-bytebuffer/src/main/scala/fluence/merkle/BinaryMerkleTree.scala]: Please add doc comment

File: statemachine/merkelized-bytebuffer/src/main/scala/fluence/merkle/ops/MerkleOperations.scala [hunk 14]
```diff
@@ -0,0 +1,26 @@
+package fluence.merkle.ops
+
+trait MerkleOperations[T] {
```
Comment (alari) [inline statemachine/merkelized-bytebuffer/src/main/scala/fluence/merkle/ops/MerkleOperations.scala]: Docs?

File: statemachine/merkelized-bytebuffer/src/main/scala/fluence/merkle/ops/MerkleOperations.scala [hunk 15]
```diff
@@ -0,0 +1,26 @@
...
+  /**
+   * Calculates hash.
+   *
+   * @param t value to hash
+   * @return hash
+   */
+  def hash(t: T): T
+
+  /**
+   * Concatenates two values.
```
Comment (alari) [inline statemachine/merkelized-bytebuffer/src/main/scala/fluence/merkle/ops/MerkleOperations.scala]: I don't understand from the docs what it does actually.

File: statemachine/merkelized-bytebuffer/src/main/scala/fluence/merkle/storage/Storage.scala [hunk 16]
```diff
@@ -0,0 +1,8 @@
+package fluence.merkle.storage
+
+import java.util
+
+trait Storage[Element] {
```
Comment (alari) [inline statemachine/merkelized-bytebuffer/src/main/scala/fluence/merkle/storage/Storage.scala]: Why do we need this layer of abstraction?

File: build.sbt [hunk 17]
```diff
@@ -118,6 +118,19 @@ lazy val `vm-hello-world-runner` = (project in file("vm/src/it/resources/test-ca
   .dependsOn(vm, `vm-hello-world`)
   .enablePlugins(AutomateHeaderPlugin)
 
+lazy val `merkelized-bytebuffer` = (project in file("statemachine/merkelized-bytebuffer"))
```
Comment (alari) [inline build.sbt]: Let's move it to `vm/merkelized-bytebuffer`.

File: statemachine/merkelized-bytebuffer/src/main/scala/fluence/merkle/ops/ByteMerkleOperations.scala [hunk 18]
```diff
@@ -0,0 +1,10 @@
+package fluence.merkle.ops
+
+class ByteMerkleOperations(hashFunc: Array[Byte] => Array[Byte]) extends MerkleOperations[Array[Byte]] {
```
Comment (alari) [inline statemachine/merkelized-bytebuffer/src/main/scala/fluence/merkle/ops/ByteMerkleOperations.scala]: Could you add docs please

File: statemachine/merkelized-bytebuffer/src/main/scala/fluence/merkle/TreeMath.scala (hunk 6 above)
Comment (michaelvoronov) [reply statemachine/merkelized-bytebuffer/src/main/scala/fluence/merkle/TreeMath.scala]: Or https://stackoverflow.com/a/3305710

Comment (DieMyst) [reply statemachine/merkelized-bytebuffer/src/main/scala/fluence/merkle/TreeMath.scala]: I have used `log2nlz` from this link already.

File: vm/merkelized-bytebuffer/src/main/scala/fluence/merkle/BinaryMerkleTree.scala [hunk 19]
```diff
@@ -0,0 +1,306 @@
...
+   *
+   * @param height position of node in a 'column'
+   * @param pos position of node in a row
+   * @return
+   */
+  private def getChildren(height: Int, pos: Int): (Array[Byte], Array[Byte]) = {
+    // leafs don't have children
+    assert(height < treeHeight)
+    val index = getNodeIndex(height + 1, pos * 2)
+    if (index % 2 != 0) (allNodes(index), allNodes(index + 1))
```
Comment (folex) [inline vm/merkelized-bytebuffer/src/main/scala/fluence/merkle/BinaryMerkleTree.scala]: move `index % 2 != 0` to `val leftmost`

File: vm/merkelized-bytebuffer/src/main/scala/fluence/merkle/BinaryMerkleTree.scala [hunk 20]
```diff
@@ -0,0 +1,313 @@
...
+   */
+  def apply(
+    treeHasher: TreeHasher,
+    storage: TrackingMemoryBuffer
+  ): BinaryMerkleTree = {
+    import TreeMath._
+
+    val chunkSize = storage.chunkSize
+
+    val size = storage.capacity()
```
Comment (michaelvoronov) [inline vm/merkelized-bytebuffer/src/main/scala/fluence/merkle/BinaryMerkleTree.scala]: `size` is too generous, because here we have several different sizes. Maybe `storageSize` or smth like it?

File: vm/merkelized-bytebuffer/src/main/scala/fluence/merkle/BinaryMerkleTree.scala [hunk 21]
```diff
@@ -0,0 +1,313 @@
...
+    if (size % chunkSize != 0)
+      throw new RuntimeException(s"Size should be divided entirely on chunkSize. Size: $size, chunkSize: $chunkSize")
+
+    // count of leaves that mapped on byte array (can have non-zero values)
+    val mappedLeafCount = (size + chunkSize - 1) / chunkSize
+
+    // number of leaves in ByteBuffer
+    val leafCount = {
```
Comment (michaelvoronov) [inline vm/merkelized-bytebuffer/src/main/scala/fluence/merkle/BinaryMerkleTree.scala]: There is a common [algo](https://graphics.stanford.edu/~seander/bithacks.html#RoundUpPowerOf2) for rounding up to the next highest power of 2

File: vm/merkelized-bytebuffer/src/main/scala/fluence/merkle/BinaryMerkleTree.scala [hunk 22]
```diff
@@ -0,0 +1,313 @@
...
+      else power2(power + 1)
+    }
+
+    val treeHeight = {
+      val h = log2(leafCount)
+      if (h == 0) 1
+      else h
+    }
+
+    val numberOfNodes = power2(treeHeight + 1) - 1
```
Comment (michaelvoronov) [inline vm/merkelized-bytebuffer/src/main/scala/fluence/merkle/BinaryMerkleTree.scala]: Hmm, why not `2*leafCount - 1`?

File: vm/merkelized-bytebuffer/src/main/scala/fluence/merkle/BinaryMerkleTree.scala [hunk 23]
```diff
@@ -0,0 +1,313 @@
...
+class BinaryMerkleTree private (
+  val allNodes: Array[Array[Byte]],
+  val treeHeight: Int,
+  mappedLeafCount: Int,
+  treeHasher: TreeHasher,
+  memory: TrackingMemoryBuffer
+) {
+  import TreeMath._
+
+  val leafsCount: Int = power2(treeHeight)
```
Comment (michaelvoronov) [inline vm/merkelized-bytebuffer/src/main/scala/fluence/merkle/BinaryMerkleTree.scala]: Why not passing leafsCount as a ctor param? Now it is computed twice here and in apply method.

File: vm/merkelized-bytebuffer/src/main/scala/fluence/merkle/BinaryMerkleTree.scala [hunk 24]
```diff
@@ -0,0 +1,313 @@
...
+    import TreeMath._
+
+    val chunkSize = storage.chunkSize
+
+    val size = storage.capacity()
+    if (size % chunkSize != 0)
+      throw new RuntimeException(s"Size should be divided entirely on chunkSize. Size: $size, chunkSize: $chunkSize")
+
+    // count of leaves that mapped on byte array (can have non-zero values)
+    val mappedLeafCount = (size + chunkSize - 1) / chunkSize
```
Comment (michaelvoronov) [inline vm/merkelized-bytebuffer/src/main/scala/fluence/merkle/BinaryMerkleTree.scala]: From my point of view, `mapped leaf` could have a lot of meaning, and `backed leafs` or `non zero leafs` is more suitable for this case.

File: vm/src/main/scala/fluence/vm/wasm/MemoryHasher.scala [hunk 25]
```diff
@@ -0,0 +1,161 @@
...
+    val digester = MessageDigest.getInstance(SHA_256)
+    merkleHasher(memory, TreeHasher(digester))
+  }
+
+  /**
+   * Instantiates the class, that get all memory and hash it with SHA-256 algorithm.
+   *
+   */
+  def buildPlainHasher(memory: MemoryBuffer): MemoryHasher = {
+    val alghoritm = SHA_256
```
Comment (michaelvoronov) [inline vm/src/main/scala/fluence/vm/wasm/MemoryHasher.scala]: Again here
```
val alghoritm = SHA_256   <--- a separated variable
val leafDigester = MessageDigest.getInstance(alghoritm)
```
but in buildMerkleTreeHasher:
```
val digester = MessageDigest.getInstance(SHA_256)  <--- no separated variable
```

File: vm/merkelized-bytebuffer/src/main/scala/fluence/merkle/TrackingMemoryBuffer.scala [hunk 26]
```diff
@@ -0,0 +1,193 @@
...
+  def getDirtyChunks: util.BitSet = dirtyChunks
+
+  /**
+   * Marks all chunks untouched.
+   */
+  def clearDirtyChunks(): Unit = dirtyChunks.clear()
+
+  private def touch(index: Int, size: Int): Unit = {
+    val from = index / chunkSize
+    val to = (index + size) / chunkSize
```
[the remark this answers is not in the document]
Comment (michaelvoronov) [reply vm/merkelized-bytebuffer/src/main/scala/fluence/merkle/TrackingMemoryBuffer.scala]: Please add test for this off-by-one.

File: vm/merkelized-bytebuffer/src/main/scala/fluence/merkle/BinaryMerkleTree.scala [hunk 27]
```diff
@@ -0,0 +1,313 @@
...
+  private val startOfLeafIndex = leafsCount
+
+  private def concatenate(l: Array[Byte], r: Array[Byte]): Array[Byte] =
+    if (r == null) l
+    else l ++ r
+
+  /**
+   * Calculates node index in array.
+   *
+   * @param height position of node in a 'column'
```
Comment (michaelvoronov) [inline vm/merkelized-bytebuffer/src/main/scala/fluence/merkle/BinaryMerkleTree.scala]: I think, it is better to use terms like `level` (or `tree level`) and position (or `position on the level`) instead of column and row. Because columns and rows usually used for describing matricies.

In this case code will be more self-documented and there won't any misuderstanding with height. I mean that for me `height` is a level counts from  the leafs up to current level, bu here it is used vise versa.

```
/**
 * Calculates node index in array.
 *
 * @param treeLevel a tree level where node is located (counted from the counted from the root to bottom)
 * @param pos a position of node in this level
 */
private def getNodeIndex(treeLevel: Int, pos: Int) = {
   assert(height > 0 && height <= treeHeight)

    // compute the first node position of given tree level in array
    val levelId = power2(treeLevel)
    val nodeId = levelId + pos - 1
    nodeId
}
```

File: vm/merkelized-bytebuffer/src/main/scala/fluence/merkle/BinaryMerkleTree.scala [hunk 28]
```diff
@@ -0,0 +1,313 @@
...
+   * @param pos position of node in a row
+   * @return
+   */
+  private def getChildren(height: Int, pos: Int): (Array[Byte], Array[Byte]) = {
+    // leafs don't have children
+    assert(height < treeHeight)
+    val index = getNodeIndex(height + 1, pos * 2)
+    val leftMost = index % 2 != 0
+
+    if (leftMost) (allNodes(index), allNodes(index + 1))
```
Comment (michaelvoronov) [inline vm/merkelized-bytebuffer/src/main/scala/fluence/merkle/BinaryMerkleTree.scala]: This method is called very often. I suggest to use some bit hacks here :). You can write
```
(allNodes(index - leftMost^1), allNodes(index + leftMost^0))
```
instead this `if`. if could markedly affect of performance because of wrong branch prediction.

But it is up to you.

File: vm/merkelized-bytebuffer/src/main/scala/fluence/merkle/BinaryMerkleTree.scala [hunk 29]
```diff
@@ -0,0 +1,313 @@
...
+    assert(height > 0 && height <= treeHeight)
+    val startOfLine = power2(height)
+    val index = startOfLine + pos - 1
+    index
+  }
+
+  /**
+   * Find children by node's position.
+   *
+   * @param height position of node in a 'column'
```
Comment (michaelvoronov) [inline vm/merkelized-bytebuffer/src/main/scala/fluence/merkle/BinaryMerkleTree.scala]: the same for height, columns and rows

File: vm/merkelized-bytebuffer/src/main/scala/fluence/merkle/BinaryMerkleTree.scala [hunk 30]
```diff
@@ -0,0 +1,313 @@
...
+   * @param pos position of node in a row
+   * @return
+   */
+  private def getChildren(height: Int, pos: Int): (Array[Byte], Array[Byte]) = {
+    // leafs don't have children
+    assert(height < treeHeight)
+    val index = getNodeIndex(height + 1, pos * 2)
+    val leftMost = index % 2 != 0
```
Comment (michaelvoronov) [inline vm/merkelized-bytebuffer/src/main/scala/fluence/merkle/BinaryMerkleTree.scala]: Is it possible for getNodeIndex to return odd index if we supplied pos*2 index as a parameter for getNodeIndex?

File: vm/merkelized-bytebuffer/src/main/scala/fluence/merkle/BinaryMerkleTree.scala [hunk 31]
```diff
@@ -0,0 +1,313 @@
...
+    val hash = recalculateLeafs(mappedLeafCount, memory.getDirtyChunks)
+    memory.clearDirtyChunks()
+    hash
+  }
+
+  // for test purpose only
+  def recalculateAll(): Array[Byte] = {
+    val allLeafs = mappedLeafCount
+    val bs = new util.BitSet(allLeafs)
+    bs.set(0, allLeafs)
```
Comment (michaelvoronov) [inline vm/merkelized-bytebuffer/src/main/scala/fluence/merkle/BinaryMerkleTree.scala]: Hmm, why 0 there?

Also, is this method really needed? You can make chunks dirty in test bb wrapper in tests. Also, there are too much methods with names `calculates`/`recalcultes` and so on - it is a little bit confusing and make this code less readable and clean.

File: vm/merkelized-bytebuffer/src/main/scala/fluence/merkle/BinaryMerkleTree.scala [hunk 32]
```diff
@@ -0,0 +1,313 @@
...
+  val treeHeight: Int,
+  mappedLeafCount: Int,
+  treeHasher: TreeHasher,
+  memory: TrackingMemoryBuffer
+) {
+  import TreeMath._
+
+  val leafsCount: Int = power2(treeHeight)
+  private val chunkSize = memory.chunkSize
+  private val startOfLeafIndex = leafsCount
```
Comment (michaelvoronov) [inline vm/merkelized-bytebuffer/src/main/scala/fluence/merkle/BinaryMerkleTree.scala]: Seems that `theFirstLeafId` with some comment more readable. With `startOfLeafIndex` someone could think that we have separate leafIndex and this variable represent its start.

File: vm/merkelized-bytebuffer/src/main/scala/fluence/merkle/BinaryMerkleTree.scala [hunk 33]
```diff
@@ -0,0 +1,313 @@
...
+  }
+
+  /**
+   * Gets data from storage and calculates hash.
+   *
+   * @param pos leaf position
+   */
+  private def recalculateLeafHash(pos: Int): Unit = {
+    if (pos < mappedLeafCount) {
+      val index = startOfLeafIndex + pos - 1
```
Comment (michaelvoronov) [inline vm/merkelized-bytebuffer/src/main/scala/fluence/merkle/BinaryMerkleTree.scala]: It is better to define variables as more close to the place of their first usage as possible.

File: vm/merkelized-bytebuffer/src/main/scala/fluence/merkle/TrackingMemoryBuffer.scala [hunk 34]
```diff
@@ -0,0 +1,193 @@
...
+    bb.put(b)
+    dirtyChunks.set(index / chunkSize)
+    this
+  }
+
+  def get(index: Int): Byte = bb.get(index)
+
+  def put(index: Int, b: Byte): TrackingMemoryBuffer = {
+    bb.put(index, b)
+    getDirtyChunks.set(index / chunkSize)
```
Comment (michaelvoronov) [inline vm/merkelized-bytebuffer/src/main/scala/fluence/merkle/TrackingMemoryBuffer.scala]: the same with `dirtyChunks`

File: vm/merkelized-bytebuffer/src/main/scala/fluence/merkle/BinaryMerkleTree.scala [hunk 35]
```diff
@@ -0,0 +1,313 @@
...
+   *
+   * @return root hash
+   */
+  @scala.annotation.tailrec
+  private def recalculateNodes(rowSize: Int, height: Int, dirtyNodes: util.BitSet): Array[Byte] = {
+    var dirtyNodeId = dirtyNodes.nextSetBit(0)
+    while (dirtyNodeId >= 0 && dirtyNodeId < rowSize) {
+      calculateNodeHash(height, dirtyNodeId)
+
+      dirtyNodes.set(getParentPos(dirtyNodeId))
```
Comment (michaelvoronov) [inline vm/merkelized-bytebuffer/src/main/scala/fluence/merkle/BinaryMerkleTree.scala]: It seems that we need to clear dirtyNodeId bit here to not to do unnecessary computation on the next iteration of recalculation.

File: vm/merkelized-bytebuffer/src/main/scala/fluence/merkle/BinaryMerkleTree.scala [hunk 36]
```diff
@@ -0,0 +1,313 @@
...
+   * @param pos child position in a row
+   */
+  private def getParentPos(pos: Int) = pos / 2
+
+  /**
+   * Returns current calculated hash without any state changes.
+   *
+   */
+  def getHash: Array[Byte] = {
+    calculateRootHash()
```
Comment (michaelvoronov) [inline vm/merkelized-bytebuffer/src/main/scala/fluence/merkle/BinaryMerkleTree.scala]: What about to caching root hash?

File: vm/merkelized-bytebuffer/src/main/scala/fluence/merkle/BinaryMerkleTree.scala [hunk 37]
```diff
@@ -0,0 +1,313 @@
...
+
+  /**
+   * Calculates node index in array.
+   *
+   * @param height position of node in a 'column'
+   * @param pos position of node in a row
+   */
+  private def getNodeIndex(height: Int, pos: Int) = {
+    assert(height > 0 && height <= treeHeight)
+    val startOfLine = power2(height)
```
Comment (michaelvoronov) [inline vm/merkelized-bytebuffer/src/main/scala/fluence/merkle/BinaryMerkleTree.scala]: It seems that it is better to precompute it.

File: vm/merkelized-bytebuffer/src/main/scala/fluence/merkle/BinaryMerkleTree.scala (hunk 30 above)
Comment (DieMyst) [reply vm/merkelized-bytebuffer/src/main/scala/fluence/merkle/BinaryMerkleTree.scala]: You are right. The result will be always odd (because we have the first node, that add `-1` to results)

File: vm/merkelized-bytebuffer/src/main/scala/fluence/merkle/BinaryMerkleTree.scala (hunk 28 above)
Comment (DieMyst) [reply vm/merkelized-bytebuffer/src/main/scala/fluence/merkle/BinaryMerkleTree.scala]: Since all the results are odd, I will remove the `if` condition

File: vm/merkelized-bytebuffer/src/main/scala/fluence/merkle/BinaryMerkleTree.scala (hunk 21 above)
Comment (DieMyst) [reply vm/merkelized-bytebuffer/src/main/scala/fluence/merkle/BinaryMerkleTree.scala]: I don't know how to make it work with signed ints

File: vm/merkelized-bytebuffer/src/main/scala/fluence/merkle/BinaryMerkleTree.scala (hunk 22 above)
Comment (DieMyst) [reply vm/merkelized-bytebuffer/src/main/scala/fluence/merkle/BinaryMerkleTree.scala]: It is not working when we have 1 chunk. But I think it doesn't matter for tree initialization, so let's leave it how it is.

File: vm/merkelized-bytebuffer/src/main/scala/fluence/merkle/TrackingMemoryBuffer.scala [hunk 38]
```diff
@@ -0,0 +1,193 @@
...
+
+  /**
+   * Marks all chunks untouched.
+   */
+  def clearDirtyChunks(): Unit = dirtyChunks.clear()
+
+  private def touch(index: Int, size: Int): Unit = {
+    val from = index / chunkSize
+    val to = (index + size) / chunkSize
+    if (from == to) getDirtyChunks.set(from)
```
[the remark this answers is not in the document]
Comment (DieMyst) [reply vm/merkelized-bytebuffer/src/main/scala/fluence/merkle/TrackingMemoryBuffer.scala]: in BitSet such code:
```
if (fromIndex == toIndex)
            return;
```
